docs/proposals/promotion-strategy.md

```diff
@@ -0,0 +1,475 @@
+
+### Will Argo CD provide out-of-the-box integration with this tooling?
+
+We believe Argo CD users would benefit greatly from the features provided by this system. But it's possible that the functionality could be offered via add-ons, including adjacent controllers and UI extensions.
+
+### What SCMs will be supported by default?
+
+We believe that this tool should provide first-class support for at least GitHub and GitLab. But we do understand that there's a tradeoff between number of SCMs supported and maintenance cost.
+
+We should implement the first-class SCM support using the same plugin system that other SCMs use, so that we ensure a good plugin development/installation/use experience.
+
+### What terminology should we use for manifests and manifest sources?
+
+"Rendered manifests" has become a common term of art in the last ~2 years. But it has no obvious corrollary term for "non-rendered manifest sources."
+
+Options:
+* **dry (Don't Repeat Yourself) manifests / hydrated manifests**
+* Manifest sources / rendered manifests
+* Others?
+
+We should choose something clear and intuitive. For now, the proposal uses "dry/hydrated."
+
+### What should be the API group(s)?
+
+The CRDs defined below use `promoter.argoproj.io/v1alpha1` and `commit-status.argoproj.io/v1alpha1`.
+
+Should those be collapsed into a single group?
+
+Should these CRDs be under an `argoproj` group at all, or maybe something new?
+
+### How should we represent repo access?
+
+Argo CD's model of using Secrets to represent repositories has problems:
+
+1) No fine-grained RBAC access: you either have access to Secrets or not, there's no way to specify "just repo secrets"
+2) Argo-arbitrary interface: the Secret schema is meaningful only for Argo and not for other GitOps operators/tools
+3) No schema validation: since Secrets have a relatively open-ended spec, writing and operating on repo secrets requires special care
+
+This seems like a good opportunity to define a new repo interface which has a clear and specific API and can be easily used by other tools.
+
+### How should the system handle change contention?
+
+To keep things simple, we think new changes should always supersede old changes.
+
+Consider a PromotionStrategy covering dev, test, and prod. If the test environment has an open PR against the hydrated branch (pending change), and someone pushes another commit to the main (DRY) branch, the PR will be updated to include the new changes, and the PR will wait for the dev environment to finish deploying and verifying that new commit.
+
+It's possible that a high-churn low environment could block merges on higher environments. So far, we think that's an acceptable tradeoff. Our recommendation will be that teams move as much of the change validation as possible to the pre-commit stage so that changes can be promoted relatively quickly through environments, and high-churn low environments won't block promotions to higher environments.
```

Comment:
In order for the proposed system to be universally accepted by the community, please consider that many organizations test changes in one environment for arbitrarily long before promoting to the next one. 

Let's consider the following scenario. As far as I understand the proposal, the change to the dry branch is reviewed, approved and merged first. The change to the hydrated `dev` branch is then merged automatically as well, but the PRs to the hydrated `test` and `prod` branches remain open. After some time the user merges the PR to the hydrated `test` branch. After another few days the software is ready to be approved and merged to `prod`. But how many days have passed since reviewing the dry branch? In this hypothetical scenario it took 7 days. So the hydrated change to `prod` is being reviewed 7 days after the equivalent change to the dry branch. I find this odd. I believe the dry and hydrated changes should be approved by the same person at the same time. Especially that the dry change to one environment may not be the same as the dry change to another environment (for example, consider a change to an URL that is different in each environment). 

This is a fundamental concern I have with this proposal – the reviews, approvals and merges to the dry branch and the hydrated branches occur at completely different times. The person who approved the dry branch may not even be at work at the time the hydrated branches are being reviewed.

I find it confusing for other reasons as well: if I'm browsing the history of the changes to the dry branch, the merge time is divorced from the deployment request. That breaks one of the promises of GitOps: that `git log` roughly corresponds to the history of deployment requests.

I appreciate this proposal, however, especially the idea to make the PR the primary interface for promotions. So let me step back and consider what kind of reviews/approvals an organization may require between promotions to different environments:
1. An organization may require that the review and approval happens at the time of deployment to a given environment. For example, there may be a review and approval before deployment to `test` and another review/approval before deployment to `staging`. In this workflow, the user can have an arbitrarily long delay between deployments to `test` and `staging`.
2. An organization may require that the review and approval happens at the time of the first deployment in a  sequence of environments. For example, deployment to `staging` may require a review and approval of changes to both `staging` and `prod`, because developers should only be allowed to deploy to `staging` when they are ready to go to `prod` as well. In this workflow the software gets promoted from `staging` to `prod` within a couple of hours from deployment to staging.

Both of the workflows above are valid. So how can reviews and approvals look like in each case? Let me propose a different solution to the one you are proposing, but one that adheres to the same fundamental idea: that git and pull requests are the user interface.

In the first scenario the entire dry configuration for `test` and `staging` would be kept in separate directories in the dry branch, so that deploying to each environment involves a separate PR to the dry branch (one PR to the `test` directory of the dry branch and one PR to the `staging` directory of the dry branch). There is no need to have a PR to the rendered branch – the automation can just create a simple feature branch to be merged to the rendered branch (this also solves the PR rate limiting problem). Each PR can then be reviewed together with the corresponding rendered change at the same time as the dry change and by the same person. After merging the `test` PR, the user can wait arbitrarily long to raise the `staging` PR. The link to raise the `staging` PR is automatically generated and added as a comment to the `test` PR, so that the user can easily execute the promotion.

In the second scenario, the dry state of `staging` and `prod` should be kept in a single directory, i.e. using shared files (with the exception of some environment-specific overrides). That way a single PR produces rendered changes to both environments, which can be reviewed again at the same time as the dry change and by the same person. However, merging the PR only triggers deployment to `staging`. You might ask: how do you then approve promotions from staging to prod? By a special comment to the PR (such as `/promote`). The prod sync would wait until the user adds such a comment (this can be implemented in a number of ways, but I'll leave that discussion for another thread).

The advantages of the approaches above is that they don't separate out the reviews of the dry and hydrated manifests into steps that are done separately and a long time apart.

What do you think?

Reply:
## Responses to concerns

> I believe the dry and hydrated changes should be approved by the same person at the same time.

100% agree. I think this proposal only makes sense when coupled with an automated diff bot which comments _all_ hydrated manifest changes on the dry PR for simultaneous review. That piece of the puzzle is discussed in the [ArgoCon talk](https://www.youtube.com/watch?v=p5EPKY3vM-E) but isn't yet sufficiently represented in this proposal.

> the reviews, approvals and merges to the dry branch and the hydrated branches occur at completely different times. The person who approved the dry branch may not even be at work at the time the hydrated branches are being reviewed.

I think we could assume that, for many environments, there will be no direct human review applied to the hydrated branch PR. We'd assume that the review was already done as part of the dry PR review. Of course, a user may choose to require a human review on the hydrated PR as an extra safety check. But I don't think that would be the default mode.

> if I'm browsing the history of the changes to the dry branch, the merge time is divorced from the deployment request

I believe we would try to make it clear that the dry commit history is a history of the developers' intent and the hydrated branch commit histories are the histories of the environments' states. The hydrated branch commits will be tied back to the dry commit history via README files and metadata files which contain auto-generated references to the associated dry commits.

## Responses to alternative proposals

### Proposal 1: developer-controlled wait between promotions

I'll answer the first proposal in two parts:

#### Part 1: two dry PRs

> Let me propose a different solution... the entire dry configuration for test and staging would be kept in separate directories in the dry branch, so that deploying to each environment involves a separate PR to the dry branch

I think the proposed system already accommodates this flow. If the developer wants to explicitly push only to `stage`, wait for a time, and then only push to `prod`, they can use the proposed system. After merging the first PR, only one promotion (hydrated) PR will be opened: one for the affected `stage` branch. After some time, the developer can open a second PR affecting only the `prod` environment, merge it, and a hydrated PR will be automatically opened/merged for the prod environment. Of course, making dry changes that affect all environments and then relying on automation to promote that change will be the preferred flow. But there is no technical limitation preventing them from taking a more manual approach. (The question of whether hydrated changes are represented as PRs or as direct commits is discussed in a separate thread, so I'll leave that discussion there.)

#### Part 2: auto-generating PR links

> The link to raise the staging PR is automatically generated and added as a comment to the test PR, so that the user can easily execute the promotion.

The link would open a PR against the stage directory of the dry branch or the stage hydrated branch? If the former, how would the system know what the source branch of the dry change is? If the latter, I feel like that's effectively the same as requiring a manual review on the auto-generated stage promotion PR against the hydrated branch.

### Proposal 2: `/promote`-based promotion PRs

I'm not sure how requiring a `/promote` comment on the dry PR is effectively different from requiring a manual approval on the hydrated PR for the target environment (via branch protection rules).

> The advantages of the approaches above is that they don't separate out the reviews of the dry and hydrated manifests into steps that are done separately and a long time apart.

I think both of the proposals above assume there is no automated diff presented to the developer on the dry PR. If I'm correct, we might not need to dive into the finer points of the alternative proposals, because your concerns may be resolved by the existence of the diff bot. :-)

Reply:
Thanks for the thorough response :-) My comments are inline:

> > Proposal 1: developer-controlled wait between promotions
> > Part 2: auto-generating PR links
> > The link to raise the staging PR is automatically generated and added as a comment to the test PR, so that the user can easily execute the promotion.

> The link would open a PR against the stage directory of the dry branch or the stage hydrated branch? If the former, how would the system know what the source branch of the dry change is? If the latter, I feel like that's effectively the same as requiring a manual review on the auto-generated stage promotion PR against the hydrated branch.

It's the former – the link would open a PR against the `stage` directory of the dry branch. The system operates on the PR to the `test` environment, so it knows the branch associated with the PR. 

There are actually some nuances here. There are two use cases:
1. The developer deploys a number of changes to `test`, as they iteratively develop the change. So a number of PRs have already been deployed to `test`. Then the developer wants to deploy all of those changes to `staging`, so they effectively want to `copy` the current state of `test` to `staging`.
2. The developer is testing the newer version of the software in `test`, while production is running an older version. However, in the meantime, the developer needs to change the configuration in all environments: for example, they need to change a URL in the config map. This URL needs to change independently from the container image version running in each environment. So the developer raises a PR to `test` (the PR contains only the change to the URL) and effectively wants to apply the same patch to `staging` without applying all the other changes from `test` to `staging`. 

We have been discussing use case 1, but use case 2 requires additional discussion. 

Back when people used different branches for each environment, they could use the `git cherry-pick` command to perform the promotion in use case 2. But nowadays we use directories per environment (for good reasons), so we need an equivalent operation.

This is why I'm thinking about producing two links per environment for the user: one that raises a PR with the `copy/replace` semantics and the other with the `cherry-pick` semantics. The automation would retrieve the list of available environments, push two feature branches per environment and generate links in the form of: 
https://github.com/my_org/gitops-dry/pull/new/$BRANCH

> > Proposal 2: /promote-based promotion PRs

> I'm not sure how requiring a /promote comment on the dry PR is effectively different from requiring a manual approval on the hydrated PR for the target environment (via branch protection rules).

You are correct that the workflow is practically the same, but there are some implementational differences:

1. I believe it's a little confusing that the action of promoting from `staging` to `prod` involves approving a PR, because, as we previously discussed, the review of that PR actually gets done earlier in the process.
2. At the organization I'm currently contracting to, the review of the dry and rendered manifests is done by a developer, whereas the decision to go from `staging` to `prod` is made by a QA engineer. These are different people and I would rather avoid the confusion that the QA engineer supposedly reviews the rendered manifests, as this is how it's going to look like in the GitHub PR record.
3. There is also the question of access control: the QA engineer should only be granted the right to push a button to approve the promotion from `staging` to `prod`. Granting them the rights to approve rendered branches is going to look confusing and this is particularly important for organizations in regulated environments, which have to explain to auditors why certain user groups have certain rights. On the other hand, the access control of who can use the `/promote` comment can be implemented in the action that processes the comment and can be limited to that action only.
4. As discussed previously, using `/promote` avoids the PR creation rate limit problem. I'm not concerned about exceeding the rate limit when creating the PRs to the dry branch, because they are typically raised by humans and the limit applies to a user. It's creating PRs in the rendered branch that has the potential to exceed the limit, because these PRs would get raised by a bot. 

Finally, I'd like to note that there are potentially other ideas than `/promote`. @Oded-B suggested to me sometime ago that it might be possible to use Markdown checklists:

- [ ] promote to prod

Checking the box is an event that can trigger an action. However, I haven't tried this method myself and I haven't tested if the event contains the username of the user who checked/unchecked the box and if the action of checking and unchecking the box is properly recorded for auditing purposes. This would require some more investigation.

Reply:
> so they effectively want to copy the current state of test to staging.

How does the promotion system know what to copy in the dry branch? Consider the example change "convert Deployment to Argo Rollouts." For dev, you might want a very simple Rollout with just a couple steps. For test, you might want something slightly more cautious. And then for prod, you might have a very cautious Rollout configuration with many steps. The promotion system has no concept of a "change" that it can map to the dry manifests, so it cannot function by copying yaml from one environment directory to the other. It's up to the developer to express their "change" as yaml in all environment directories (or, if they choose, one at a time for extra control), and then it's up to the promotion system to deploy those changes according to the defined promotion rules. I'm not sure how we could implement a generalizable promotion system based on copying dry YAML from one environment directory to the other. (Codefresh's promotion system uses what amount to YAML "masks," implemented as json path expressions; I think that system is fraught with edge cases.)

> I believe it's a little confusing that the action of promoting from staging to prod involves approving a PR

It doesn't _have_ to involve approving a PR. You could implement a custom commit status check requiring that all prod promotions must be approved in some internal QA system. The status of that check would appear on the PR in the list of checks, and you could could add a "details" link out to your QA approval system for discoverability.

> These are different people and I would rather avoid the confusion that the QA engineer supposedly reviews the rendered manifests, as this is how it's going to look like in the GitHub PR record.

Understandable. I think the commit status checks-based solution resolves that concern.

> using /promote avoids the PR creation rate limit problem

Personally I think direct-commit is a better way to avoid the rate limit issue. But one could still use a `/promote` mechanism on rendered manifests instead of dry manifests.